TESTS/netsocket/udp/udpsocket_sendto_repeat.cpp

```diff
@@ -37,18 +37,13 @@ void UDPSOCKET_SENDTO_REPEAT()
     Timer timer;
     int i;
     static const char tx_buffer[] = {'h', 'e', 'l', 'l', 'o'};
-    bool oom_earlier = false; // 2 times in a row -> time to give up
     for (i = 0; i < 100; i++) {
         sent = sock.sendto(udp_addr, tx_buffer, sizeof(tx_buffer));
-        if (sent == NSAPI_ERROR_NO_MEMORY) {
-            if (oom_earlier) {
-                break;
-            }
-            oom_earlier = true;
+        if (sent == NSAPI_ERROR_NO_MEMORY || sent == NSAPI_ERROR_BUSY) {
```

Comment:
How is BUSY allowed in `sendto()` call?

What is the difference of WOULD_BLOCK?

Reply:
`When a socket is in **nonblocking** mode, socket operations return NSAPI_ERROR_WOULD_BLOCK when a transaction cannot immediately complete..`

Can we return WOULD_BLOCK also for blocking sockets and also without calling sigio callbacks?

Reply:
Drivers and stack are always in non-blocking mode.
`TCPSocket` and `UDPSocket` are the modules which handle the blocking operations.

See https://os.mbed.com/docs/mbed-os/v5.12/porting/networkstack.html and a tittle `Providing the Socket API`

Reply:
@SeppoTakalo So it's fine to just change BUSY to WOULD_BLOCK at the BC95 modem driver and Greentea test case?

Reply:
Returning "WOULD_BLOCK" is non-conventional for UDP sockets - it is more normal to just drop the packet and treat as congestion with something like "BUSY". But I believe we've seen cases where the behaviour of the device meant it was the least bad option in practice.

If you do return WOULD_BLOCK, you must at some point generate a callback indicating when to retry - the blocking layer above depends on this.

Here, I'd be inclined to say that the test should tolerate "BUSY" as a possible error, but not sure I've got all info.

Reply:
One performance issue is DTLS - which likes to send 5 packets back-to-back in a burst. (Although it may have been improved). A device that reliably drops packets 2-5 in that situation would cause performance issues - returning "WOULD_BLOCK" causes a blocking socket to wait and retry on packets 2-5.

Generally, sending UDP protocols should be coded to handle congestion anywhere on the path, including locally, but if the local buffers are particularly small, causing instant congestion on a small burst, that may be worse than the protocol can handle, so having internal "flow control" with the "would block" return helps smooth it out.

That is a good way to think about "would block" - local flow control.